[inductor] handle offset in ReinterpretView for alignment

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151859
* #151841

Fix https://github.com/pytorch/pytorch/issues/151589 

It's interesting that the Q4_K dequantization example in the referred GH issue does not crash even if Inductor pass triton the wrong alignment information. I dig this a bit. The main reason is, there are 2 things in triton that decides the vectorization size
1. alignement
2. max number of contiguous elements a thread need to process

Here is the triton code that decides vectorization size [link](https://github.com/triton-lang/triton/blob/c5fed8e1ca66c1e04b773ffa9d4e49d481fc43ad/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp#L147-L157), and here is the triton code that considers contiguity for vectorization [link](https://github.com/triton-lang/triton/blob/c5fed8e1ca66c1e04b773ffa9d4e49d481fc43ad/lib/Analysis/AxisInfo.cpp#L1250-L1269)

When Inductor wrongly tell triton that a unaligned tensor is aligned, Triton may not do vectorization (or not do full vectorization) because of the second restriction.

Check this test:
```
    @parametrize(
        "size",
        (
            128,
            1024,
            1024 * 1024,
        ),
    )
    def test_slice_view_dtype(self, size):
        offset = 1

        def f(x):
            return x[2:].view(dtype=torch.float32) + 1

        x = torch.randn((size + offset) * 2, dtype=torch.bfloat16, device=self.device)
        self.common(f, (x,), reference_in_float=False)
```

Before the fix, Inductor would tell Triton that the output of aten.view.dtype tensor is aligned even though it's not. That tensor will be passed to the triton kernel for the aten.add. Triton may do different vectorization decision depending on the tensor size
1. when size = 128, triton pick ld.global.b32 to load data from global memory
2. when size = 1024, triton uses ld.global.v2.b32
4. when size = 1024 * 1024, triton uses ld.global.v4.b32

So whether wrong alignment metadata causes issue depends on if triton picks the vectorized instructions. The latter depends on the triton config (block size) decided by inductor and triton internal logic (how they assign elements to each thread). We'd better to make sure Inductor always generate correct metadata to make sure such hidden issues does not turn into crash later.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

shunting314 committed:
[inductor] handle offset in ReinterpretView for alignment

[ghstack-poisoned]

shunting314 commented:
cc @htyu to confirm if my understanding of triton vectorization is correct in the PR summary

htyu commented:
> cc @htyu to confirm if my understanding of triton vectorization is correct in the PR summary

Looks good to me. When there isn't enough work for a thread, vectorization may not happen.

Skylion007 commented on torch/_inductor/ir.py:
```diff
@@ -2595,6 +2597,29 @@ def is_stride_order_storage_and_layout(
 
 
+def is_unaligned(node: IRNode) -> bool:
+    if isinstance(node, TensorBox):
```
This and the next if statement could merged and the isinstance check could be combined?

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2595,6 +2597,29 @@ def is_stride_order_storage_and_layout(
+        return is_unaligned(node.data)
+
+    if isinstance(node, ReinterpretView):
+        has_unaligned_layout = False
```
nit: this is unused

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2595,6 +2597,29 @@ def is_stride_order_storage_and_layout(
+        has_unaligned_layout = False
+
+        layout = node.layout
+        has_unaligned_layout = not statically_known_true(
+            layout.offset * get_dtype_size(layout.dtype) % GPU_ALIGN_BYTES == 0
+        )
```
nit: this helper exists elsewhere i believe..

shunting314 commented on torch/_inductor/ir.py:
```diff
@@ -2595,6 +2597,29 @@ def is_stride_order_storage_and_layout(
+        has_unaligned_layout = not statically_known_true(
+            layout.offset * get_dtype_size(layout.dtype) % GPU_ALIGN_BYTES == 0
+        )
+        return is_unaligned(node.data) or has_unaligned_layout
```
A ReinterpretView can be unaligned either because the underlying buffer is unaligned or because the extra offset from the layout is unaligned. The former can happen e.g. when the tensor is returned by a fallback kernel

shunting314 committed:
Update on "[inductor] handle offset in ReinterpretView for alignment"


Fix https://github.com/pytorch/pytorch/issues/151589 

It's interesting that the Q4_K dequantization example in the referred GH issue does not crash even if Inductor pass triton the wrong alignment information. I dig this a bit. The main reason is, there are 2 things in triton that decides the vectorization size
1. alignement
2. max number of contiguous elements a thread need to process

Here is the triton code that decides vectorization size [link](https://github.com/triton-lang/triton/blob/c5fed8e1ca66c1e04b773ffa9d4e49d481fc43ad/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp#L147-L157), and here is the triton code that considers contiguity for vectorization [link](https://github.com/triton-lang/triton/blob/c5fed8e1ca66c1e04b773ffa9d4e49d481fc43ad/lib/Analysis/AxisInfo.cpp#L1250-L1269)

When Inductor wrongly tell triton that a unaligned tensor is aligned, Triton may not do vectorization (or not do full vectorization) because of the second restriction.

Check this test:
```
    parametrize(
        "size",
        (
            128,
            1024,
            1024 * 1024,
        ),
    )
    def test_slice_view_dtype(self, size):
        offset = 1

        def f(x):
            return x[2:].view(dtype=torch.float32) + 1

        x = torch.randn((size + offset) * 2, dtype=torch.bfloat16, device=self.device)
        self.common(f, (x,), reference_in_float=False)
```

Before the fix, Inductor would tell Triton that the output of aten.view.dtype tensor is aligned even though it's not. That tensor will be passed to the triton kernel for the aten.add. Triton may do different vectorization decision depending on the tensor size
1. when size = 128, triton pick ld.global.b32 to load data from global memory
2. when size = 1024, triton uses ld.global.v2.b32
4. when size = 1024 * 1024, triton uses ld.global.v4.b32

So whether wrong alignment metadata causes issue depends on if triton picks the vectorized instructions. The latter depends on the triton config (block size) decided by inductor and triton internal logic (how they assign elements to each thread). We'd better to make sure Inductor always generate correct metadata to make sure such hidden issues does not turn into crash later.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov 

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge
